swift/extractor/infra/TargetFile.h

```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <string>
+#include <fstream>
+#include <iostream>
+#include <cstdio>
+#include <cerrno>
+#include <system_error>
+#include <sstream>
+
+namespace codeql {
+
+// Only the first process trying to open an `TargetFile` for a given `target` is allowed to do
+// so, all others will have an instance with `good() == false` and failing on any other operation.
+// The content streamed to the `TargetFile` is written to `workingDir/target`, and is moved onto
+// `targetDir/target` only when `commit()` is called
+class TargetFile {
+  std::string workingPath;
+  std::string targetPath;
+  std::ofstream out;
+
+ public:
+  TargetFile(std::string_view target, std::string_view targetDir, std::string_view workingDir);
+
+  bool good() const;
```

Comment:
I think we can simplify the API:
1. Replace `commit()` with a destructor. This:
   * means not having to remember to call `commit()` and
   * makes it impossible to keep using a committed file.
2. Don't make callers check `good()` (and when we use it internally, it _might_ become clearer to just inline it). Instead make goodness an invariant (from the user's perspective):
   * Make the constructor private, and add a public factory function that returns an `optional`, with the invariant that the factory either returns a `TargetFile` with an opened file or `nullopt`. This makes it clear that the caller should check for presence.
   * Replacing `commit()` with a destructor means that once you have a `TargetFile` with an open file, that file stays open as long as the object does (RAII).
   * For other, unexpected failures, we can (and already do) handle them in `TargetFile` (by logging an error and aborting).

(As always, my review comments are open to discussion if you disagree with the suggestion!)

Reply:
I agree with the factory function, but while I initially went with the destructor approach, I went for the explicit `commit` because it is what we actually want (for now). The temporary trap file was introduced because extractor crashes were causing problems: I don't remember if this was related to the known issue of the trap importer not liking incomplete trap files, or if we just did not want to have an inconsitent DB. In any case, we wanted the trap to be dumped only on successful execution, which would not be the case were we to commit on destruction.

We can definitely revisit that, but maybe in a separate PR, I'd keep that behaviour in this one. In any case, we should wait for @AlexDenisov because he might know more about this.

Reply:
> if this was related to the known issue of the trap importer not liking incomplete trap files

I believe it was that. I won't block the PR on it, but regarding:

> In any case, we wanted the trap to be dumped only on successful execution, which would not be the case were we to commit on destruction.

`std::abort` [does not call](https://en.cppreference.com/w/cpp/utility/program/abort) destructors (and obviously a crash doesn't either). Is there a scenario where we don't finish successfully, but do invoke the destructor?

> I agree with the factory function

I think this it's worth using a factory function and hiding `good()` from users regardless of the `commit()` changes. Does your agreement mean you plan to make that change in this PR?

Reply:
> `std::abort` [does not call](https://en.cppreference.com/w/cpp/utility/program/abort) destructors (and obviously a crash doesn't either). Is there a scenario where we don't finish successfully, but do invoke the destructor?

uh, I honestly thought it did, should have double checked :slightly_smiling_face: And `assert` which we currently use also calls `std::abort`, so I agree we can move `commit` to the destructor.

> Does your agreement mean you plan to make that change in this PR?

Yes :slightly_smiling_face:

Reply:
done :+1: